torch/csrc/profiler/python/combined_traceback.cpp

```diff
@@ -0,0 +1,123 @@
+#include <torch/csrc/profiler/python/combined_traceback.h>
+#include <torch/csrc/python_headers.h>
+#include <torch/csrc/utils/pybind.h>
+#include <torch/csrc/utils/pythoncapi_compat.h>
+#include <iostream>
+
+namespace py = pybind11;
+
+namespace torch {
+// Locking:
+// We need to free PyCodeObjects when ~StackContext runs, but
+// CUDACachingAllocator may hold its device lock when ~StackContext runs.
+
+// Because the thread calling the allocator _may_ hold the GIL,
+// attempting to lock the GIL in ~StackContext can deadlock:
+// T0: GIL Lock -> Call Allocator    ->| Waiting Device Lock
+// T1: Call Allocator -> Device Lock ->| Waiting GIL Lock
+// Instead the destructor defers freeing stack frames by putting them in
+// to_free_frames. We still need a lock to manage this vector, but
+// we can ensure an overall lock ordering of GIL -> device_lock ->
+// to_free_frames_mutex because ::gather is called outside of the device lock.
+
+namespace {
+static std::mutex to_free_frames_mutex;
+static std::vector<CapturedTraceback::PyFrame> to_free_frames;
+struct PythonTraceback : public CapturedTraceback::Python {
+  std::vector<CapturedTraceback::PyFrame> gather() override {
+    std::vector<CapturedTraceback::PyFrame> frames;
+    py::gil_scoped_acquire acquire;
+    {
+      std::lock_guard lock(to_free_frames_mutex);
+      for (CapturedTraceback::PyFrame f : to_free_frames) {
+        Py_XDECREF(f.code);
+      }
+      to_free_frames.clear();
+    }
+    PyFrameObject* f = PyEval_GetFrame();
+    Py_XINCREF(f);
+    while (f) {
+      frames.emplace_back(
+          CapturedTraceback::PyFrame{PyFrame_GetCode(f), PyFrame_GetLasti(f)});
+      auto f_back = PyFrame_GetBack(f);
+      Py_XDECREF(f);
+      f = f_back;
+    }
+    return frames;
+  }
+  void release(std::vector<CapturedTraceback::PyFrame>& frames) override {
+    std::lock_guard lock(to_free_frames_mutex);
+    to_free_frames.insert(to_free_frames.end(), frames.begin(), frames.end());
+  }
+  void appendSymbolized(
+      const std::vector<CapturedTraceback::PyFrame>& to_symbolize,
+      SymbolizedTracebacks& result) override {
+    auto torch = py::module::import("torch");
```

Comment:
hmm?

Reply:
oh, that is suppose to be part of the next PR, split them wrong.